### PR TITLE
Ignore `No such key` exceptions in some cases

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

The exception can be printed internally in some parts of code but its rethrown and handled correctly later on (e.g. ignored if it's safe to ignore)

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 5a92eb69f14e26ff8b2fb02c036777a531d41ad4 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71236/5a92eb69f14e26ff8b2fb02c036777a531d41ad4/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2c4d93f5523da36e4e7fb491a4e338a7c59e4d28/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by antaljanosbenjamin — APPROVED** [review]

File: tests/docker_scripts/stress_tests.lib [hunk 1]
```diff
@@ -263,8 +263,12 @@ function check_logs_for_critical_errors()
...
     [ -s /test_output/logical_errors.txt ] || rm /test_output/logical_errors.txt
 
-    # No such key errors (ignore a.myext which is used in 02724_database_s3.sh and does not exist)
-    rg --text "Code: 499.*The specified key does not exist" /var/log/clickhouse-server/clickhouse-server*.log | grep -v "a.myext" > /test_output/no_such_key_errors.txt \
+    # ignore:
+    #  - a.myext which is used in 02724_database_s3.sh and does not exist
+    #  - "DistributedCacheTCPHandler" and "caller id: None:DistribCache" because they happen inside distributed cache server
+    #  - "Will read from object storage directly" printed internally by ReadBufferFromDistributedCache, exception will be rethrown and handled correctly
+    #  - "Caught exception while reading S3 object" printed internally by ReadBufferFromS3, exception will be rethrown and handled correctly
+    rg --text "Code: 499.*The specified key does not exist" /var/log/clickhouse-server/clickhouse-server*.log | grep -v -e "a.myext" -e "DistributedCacheTCPHandler" -e "Will read from object storage directly" -e "Caught exception while reading S3 object" -e "caller id: None:DistribCache" > /test_output/no_such_key_errors.txt \
```
Comment (kssenii) [inline tests/docker_scripts/stress_tests.lib:271]: Maybe substitute "Will read from object storage directly" with logger name "ReadBufferFromDistributedCache" to make it more robust?

Comment (antonio2368) [inline tests/docker_scripts/stress_tests.lib:271]: Maybe do the same with ReadBufferFromS3 also :thinking:
I expect that all read buffers should forward such errors at some point

Comment (antonio2368) [inline tests/docker_scripts/stress_tests.lib:271]:
```suggestion
    rg --text "Code: 499.*The specified key does not exist" /var/log/clickhouse-server/clickhouse-server*.log | grep -v -e "a.myext" -e "DistributedCacheTCPHandler" -e "ReadBufferFromDistributedCache" -e "ReadBufferFromS3" -e "caller id: None:DistribCache" > /test_output/no_such_key_errors.txt \
```

Comment (kssenii) [inline tests/docker_scripts/stress_tests.lib:271]: > "caller id: None:DistribCache"

I think this one should be covered by logger names ("DistributedCacheTCPHandler", ""ReadBufferFromDistributedCache""), as it is only added to exception message here
https://github.com/ClickHouse/ClickHouse/blob/8f7fc5e49fd683c4aad0d883a1f430a9b081dfdb/src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp#L807

Comment (antonio2368) [reply tests/docker_scripts/stress_tests.lib:271]: Maybe do the same with ReadBufferFromS3 also 🤔
I expect that all read buffers should forward such errors at some point

Comment (kssenii) [reply tests/docker_scripts/stress_tests.lib:271]: Agree

Comment (antonio2368) [reply tests/docker_scripts/stress_tests.lib:271]: The exception so rethrown from the handler again, so it's printed by ServerErrorHandler again with a different logger name

File: tests/docker_scripts/stress_tests.lib [hunk 2]
```diff
@@ -263,8 +263,12 @@ function check_logs_for_critical_errors()
...
     [ -s /test_output/logical_errors.txt ] || rm /test_output/logical_errors.txt
 
-    # No such key errors (ignore a.myext which is used in 02724_database_s3.sh and does not exist)
-    rg --text "Code: 499.*The specified key does not exist" /var/log/clickhouse-server/clickhouse-server*.log | grep -v "a.myext" > /test_output/no_such_key_errors.txt \
+    # ignore:
+    #  - a.myext which is used in 02724_database_s3.sh and does not exist
+    #  - "DistributedCacheTCPHandler" and "caller id: None:DistribCache" because they happen inside distributed cache server
+    #  - "ReadBufferFromDistributedCache" exception printed internally by ReadBufferFromDistributedCache, exception will be rethrown and handled correctly
+    #  - "ReadBufferFromS3" exception printed internally by ReadBufferFromS3, exception will be rethrown and handled correctly
+    rg --text "Code: 499.*The specified key does not exist" /var/log/clickhouse-server/clickhouse-server*.log | grep -v -e "a.myext" -e "DistributedCacheTCPHandler" -e "ReadBufferFromDistributedCache" -e "ReadBufferFromS3" -e "caller id: None:DistribCache" > /test_output/no_such_key_errors.txt \
```
Comment (kssenii) [inline tests/docker_scripts/stress_tests.lib:271]: Also `ReadBufferFromAzureBlobStorage`?

Comment (kssenii) [reply tests/docker_scripts/stress_tests.lib:271]: Potentially the exception can also be logged here https://github.com/ClickHouse/ClickHouse/blob/12718ff0247365cd06f72e44b5d6b2a550276c4d/src/Disks/IO/AsynchronousBoundedReadBuffer.cpp#L360-L370
So probably change `tryLogCurrentException(__PRETTY_FUNCTION__)` -> `tryLogCurrentException(log)` and add `AsynchronousBoundedReadBuffer` to the list? Though here we indeed do not rethrow for obvious reasons, so may be questionable that we should ignore it or not.

**Review by kssenii — APPROVED** [review]

**Pull request merged**